src/zabapgit_2fa.prog.abap

```diff
@@ -0,0 +1,726 @@
+    CLASS-METHODS:
+      set_new_token_request IMPORTING ii_request   TYPE REF TO if_http_request,
+      get_token_from_response IMPORTING ii_response     TYPE REF TO if_http_response
+                              RETURNING VALUE(rv_token) TYPE string,
+      parse_repo_from_url IMPORTING iv_url              TYPE string
+                          RETURNING VALUE(rv_repo_name) TYPE string,
+      set_list_token_request IMPORTING ii_request TYPE REF TO if_http_request,
+      get_tobedel_tokens_from_resp IMPORTING ii_response   TYPE REF TO if_http_response
+                                   RETURNING VALUE(rt_ids) TYPE stringtab,
+      set_del_token_request IMPORTING ii_request  TYPE REF TO if_http_request
+                                      iv_token_id TYPE string,
+      get_authenticated_client IMPORTING iv_username      TYPE string
+                                         iv_password      TYPE string
+                                         iv_2fa_token     TYPE string
+                               RETURNING VALUE(ri_client) TYPE REF TO if_http_client
+                               RAISING   lcx_2fa_auth_failed
+                                         lcx_2fa_communication_error.
+ENDCLASS.
+
+CLASS lcl_2fa_github_authenticator IMPLEMENTATION.
+  METHOD constructor.
+    super->constructor( '^https?://(www\.)?github.com.*$' ).
+  ENDMETHOD.
+
+  METHOD authenticate.
+    DATA: li_http_client           TYPE REF TO if_http_client,
+          lv_http_code             TYPE i,
+          lv_http_code_description TYPE string.
+
+    " 1. Try to login to GitHub API
+    li_http_client = get_authenticated_client( iv_username  = iv_username
+                                               iv_password  = iv_password
+                                               iv_2fa_token = iv_2fa_token ).
+
+    " 2. Create an access token which can be used instead of a password
+    " https://developer.github.com/v3/oauth_authorizations/#create-a-new-authorization
+
+    set_new_token_request( ii_request = li_http_client->request ).
+
+    li_http_client->send( EXCEPTIONS OTHERS = 1 ).
+    IF sy-subrc <> 0.
+      raise_comm_error_from_sy( ).
+    ENDIF.
+
+    li_http_client->receive( EXCEPTIONS OTHERS = 1 ).
+    IF sy-subrc <> 0.
+      raise_comm_error_from_sy( ).
+    ENDIF.
+
+    li_http_client->response->get_status(
+      IMPORTING
+        code   = lv_http_code
+        reason = lv_http_code_description ).
+    IF lv_http_code <> 201.
+      RAISE EXCEPTION TYPE lcx_2fa_token_gen_failed
+        EXPORTING
+          iv_error_text = |Token generation failed: { lv_http_code } { lv_http_code_description }|.
+    ENDIF.
+
+    rv_access_token = get_token_from_response( li_http_client->response ).
+    IF rv_access_token IS INITIAL.
+      RAISE EXCEPTION TYPE lcx_2fa_token_gen_failed
+        EXPORTING
+          iv_error_text = 'Token generation failed: parser error' ##NO_TEXT.
+    ENDIF.
```

Comment:
Interesting that GitHub returns the token before it is fully persisted and ready to use. I also don't have that problem on my end.
I'd prefer adding `RZL_SLEEP` instead of `WAIT` to avoid the implicit commit.